bench/bench.go

```diff
@@ -0,0 +1,335 @@
+)
+
+// A Sample for a particular client
+type Sample struct {
+	JobMsgCnt int
+	MsgCnt    uint64
+	MsgBytes  uint64
+	IOBytes   uint64
+	Start     time.Time
+	End       time.Time
+}
+
+// SampleGroup for a number of samples, the group is a Sample itself agregating the values the Samples
+type SampleGroup struct {
+	Sample
+	Samples []*Sample
+}
+
+// Benchmark to hold the various Samples organized by publishers and subscribers
+type Benchmark struct {
+	Sample
+	RunID      string
+	Pubs       *SampleGroup
+	Subs       *SampleGroup
+	subChannel chan *Sample
+	pubChannel chan *Sample
+}
+
+// NewBenchmark initializes a Benchmark. After creating a bench call AddSubSample/AddPubSample.
+// When done collecting samples, call EndBenchmark
+func NewBenchmark(subCnt, pubCnt int) *Benchmark {
+	bm := Benchmark{RunID: nuid.Next()}
+	bm.Subs = NewSampleGroup()
+	bm.Pubs = NewSampleGroup()
+	if subCnt > 0 {
+		bm.subChannel = make(chan *Sample, subCnt)
+	}
+	if pubCnt > 0 {
+		bm.pubChannel = make(chan *Sample, pubCnt)
+	}
+	return &bm
+}
+
+// Close organizes collected Samples and calculates aggregates. After Close(), no more samples can be added.
+func (bm *Benchmark) Close() {
+	close(bm.subChannel)
+	close(bm.pubChannel)
+
+	for s := range bm.subChannel {
+		bm.Subs.AddSample(s)
+	}
+	for s := range bm.pubChannel {
+		bm.Pubs.AddSample(s)
+	}
+
+	bm.Start = bm.Subs.Start
+	if bm.Start.After(bm.Pubs.Start) {
+		bm.Start = bm.Pubs.Start
+	}
+	bm.End = bm.Subs.End
+	if bm.End.Before(bm.Pubs.End) {
+		bm.End = bm.Pubs.End
+	}
+
+	bm.MsgBytes = bm.Pubs.MsgBytes + bm.Subs.MsgBytes
+	bm.IOBytes = bm.Pubs.IOBytes + bm.Subs.IOBytes
+	bm.MsgCnt = bm.Pubs.MsgCnt + bm.Subs.MsgCnt
+	bm.JobMsgCnt = bm.Pubs.JobMsgCnt + bm.Subs.JobMsgCnt
+}
+
+// AddSubSample to the benchmark
+func (bm *Benchmark) AddSubSample(s *Sample) {
+	bm.subChannel <- s
+}
+
+// AddPubSample to the benchmark
+func (bm *Benchmark) AddPubSample(s *Sample) {
+	bm.pubChannel <- s
+}
+
+// CSV generates a csv report of all the samples collected
+func (bm *Benchmark) CSV() string {
+	var buffer bytes.Buffer
+	writer := csv.NewWriter(&buffer)
+	writer.Write([]string{"#RunID", "ClientID", "MsgCount", "MsgBytes", "MsgsPerSec", "BytesPerSec", "DurationSecs"})
+	groups := []*SampleGroup{bm.Subs, bm.Pubs}
+	pre := "S"
+	for i, g := range groups {
+		if i == 1 {
+			pre = "P"
+		}
+		for j, c := range g.Samples {
+			r := []string{bm.RunID, fmt.Sprintf("%s%d", pre, j), fmt.Sprintf("%d", c.MsgCnt), fmt.Sprintf("%d", c.MsgBytes), fmt.Sprintf("%d", c.Rate()), fmt.Sprintf("%f", c.Throughput()), fmt.Sprintf("%f", c.Duration().Seconds())}
+			err := writer.Write(r)
+			if err != nil {
+				log.Fatalf("Error while serializing %v: %v", c, err)
+			}
+		}
+	}
+
+	writer.Flush()
+	return buffer.String()
+}
+
+// NewSample creates a new Sample initialized to the provided values. The nats.Conn information captured
+func NewSample(jobCount int, msgSize int, start, end time.Time, nc *nats.Conn) *Sample {
```

Comment:
Although I don't see it as a problem, if bringing this dependency proves to be problematic, you would have to pass the In/Out stats values as parameters here...